network/gossip/libp2p/topology/statefulTopologyManager_test.go

```diff
@@ -0,0 +1,172 @@
+	suite.multiSystemEndToEndConnectedness(1, 20, 200, 240, 10, 200, 8)
+}
+
+// TestHighScale creates systems with
+// 40 access nodes
+// 400 collection nodes in 16 clusters
+// 480 consensus nodes
+// 20 execution nodes
+// 400 verification nodes
+// and builds a stateful topology for the systems.
+// For each system, it then checks the end-to-end connectedness of the topology graph.
+func (suite *StatefulTopologyTestSuite) TestHighScale() {
+	suite.multiSystemEndToEndConnectedness(1, 40, 400, 480, 20, 400, 16)
+}
+
+// generateSystem is a test helper that given number of nodes per role as well as desire number of clusters
+// generates the protocol state, identity list and subscription managers for the nodes.
+// - acc: number of access nodes
+// - col: number of collection nodes
+// - exe: number of execution nodes
+// - ver: number of verification nodes
+// - cluster: number of clusters of collection nodes
+func (suite *StatefulTopologyTestSuite) generateSystem(acc, col, con, exe, ver, cluster int) (protocol.State,
+	flow.IdentityList,
+	[]channel.SubscriptionManager) {
+
+	collector, _ := test.GenerateIDs(suite.T(), col, test.DryRun, unittest.WithRole(flow.RoleCollection))
+	access, _ := test.GenerateIDs(suite.T(), acc, test.DryRun, unittest.WithRole(flow.RoleAccess))
+	consensus, _ := test.GenerateIDs(suite.T(), con, test.DryRun, unittest.WithRole(flow.RoleConsensus))
+	verification, _ := test.GenerateIDs(suite.T(), ver, test.DryRun, unittest.WithRole(flow.RoleVerification))
+	execution, _ := test.GenerateIDs(suite.T(), exe, test.DryRun, unittest.WithRole(flow.RoleExecution))
+
+	ids := flow.IdentityList{}
+	ids = ids.Union(collector)
+	ids = ids.Union(access)
+	ids = ids.Union(consensus)
+	ids = ids.Union(verification)
+	ids = ids.Union(execution)
+
+	// mocks state for collector nodes topology
+	state, _ := topology.CreateMockStateForCollectionNodes(suite.T(),
+		ids.Filter(filter.HasRole(flow.RoleCollection)), uint(cluster))
+
+	subMngrs := test.MockSubscriptionManager(suite.T(), ids)
+
+	return state, ids, subMngrs
+}
+
+// multiSystemEndToEndConnectedness is a test helper evaluates end-to-end connectedness of the system graph
+// over several number of systems each with specified number of nodes on each role.
+func (suite *StatefulTopologyTestSuite) multiSystemEndToEndConnectedness(system, acc, col, con, exe, ver, cluster int) {
+	// creates a histogram to keep average fanout of nodes in systems
+	aveHist := thist.NewHist(nil, fmt.Sprintf("Average fanout for %d systems", system),
+		"fit", 10, false)
+
+	for j := 0; j < system; j++ {
+		// adjacency map keeps graph component of a single channel ID
+		adjMap := make(map[flow.Identifier]flow.IdentityList)
+
+		// creates a flow system
+		state, ids, subMngrs := suite.generateSystem(acc, col, con, exe, ver, cluster)
+
+		// creates a fanout histogram for this system
+		systemHist := thist.NewHist(nil, fmt.Sprintf("System #%d fanout", j),
```

Comment:
I would keep the whole histogram code in a different test which will be skipped by default this way we keep the test that actually test things separate from code that helps analyze thing kinda like how the benchmark test are written in Go.

Reply:
Sure, going to segregate it into another test.

Reply:
fixed in https://github.com/onflow/flow-go/pull/153/commits/9e9e5a00d532356c0034094da26bb0dd7133961f as we talked.